File: go/anthropic/messages/streamout.go
package messages

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/elkpi/oxa/go/ir"
	"github.com/elkpi/oxa/go/modelmap"
)

// StreamEncoder incrementally converts an IR event stream into Anthropic
// Messages wire events (IR -> face), enforcing the INV-5 grammar. The mapping
// is near-identity: Anthropic is the block model, so each IR event maps to
// the correspondingly named wire event. Envelope fields absent from the IR
// render with the documented defaults and record no loss.
type StreamEncoder struct {
	models    modelmap.Table
	id        string
	model     string
	started   bool
	nextIndex int
	openIndex int
	blockOpen bool
	openTool  bool
	toolInput string
	toolParts []string
	deltaSeen bool
	done      bool
}

// NewStreamEncoder returns an event-stream encoder. The variadic Options match
// the package conversion functions (WithModelMap applies to the message
// envelope Model).
func NewStreamEncoder(opts ...Option) *StreamEncoder {
	o := newOptions(opts...)
	return &StreamEncoder{models: o.models}
}

// MarshalJSON keeps message_delta's envelope form free of a zero delta type and
// keeps empty input_json_delta fragments present on the wire.
func (d StreamDelta) MarshalJSON() ([]byte, error) {
	type deltaWire struct {
		Type         string  `json:"type,omitempty"`
		Text         string  `json:"text,omitempty"`
		PartialJSON  *string `json:"partial_json,omitempty"`
		StopReason   string  `json:"stop_reason,omitempty"`
		StopSequence string  `json:"stop_sequence,omitempty"`
	}
	wire := deltaWire{
		Type:         d.Type,
		Text:         d.Text,
		StopReason:   d.StopReason,
		StopSequence: d.StopSequence,
	}
	if d.Type == "input_json_delta" || d.PartialJSON != "" {
		partial := d.PartialJSON
		wire.PartialJSON = &partial
	}
	return json.Marshal(wire)
}

// Apply pushes one IR event and returns the wire events it produces. The
// mapping is near-identity except that a tool block stop may synthesize one
// full input_json_delta when no argument deltas were supplied. Out-of-grammar
// orderings are structural errors.
func (e *StreamEncoder) Apply(ev ir.Event) ([]*StreamEvent, []ir.Loss, error) {
	if ev == nil {
		return nil, nil, fmt.Errorf("anthropic: nil event")
	}
	if e.done {
		return nil, nil, fmt.Errorf("anthropic: event applied after MessageDone (%T)", ev)
	}
	switch event := ev.(type) {
	case ir.MessageStart:
		if e.started {
			return nil, nil, fmt.Errorf("anthropic: duplicate MessageStart")
		}
		e.started = true
		e.id = event.ID
		e.model = e.models.Map(event.Model)
		// Wire envelope defaults: type "message", role "assistant", empty
		// content, null stop_reason (the Go empty string), zero usage.
		return []*StreamEvent{{Type: EventTypeMessageStart, Message: &Response{
			ID:      e.id,
			Type:    TypeMessage,
			Role:    RoleAssistant,
			Model:   e.model,
			Content: []BlockWire{},
			Usage:   &UsageWire{},
		}}}, nil, nil
	case ir.ContentBlockStart:
		if !e.started || e.blockOpen {
			return nil, nil, fmt.Errorf("anthropic: ContentBlockStart out of grammar order")
		}
		if event.Index != e.nextIndex {
			return nil, nil, fmt.Errorf("anthropic: ContentBlockStart index %d, want %d", event.Index, e.nextIndex)
		}
		switch block := event.Block.(type) {
		case ir.TextBlock:
			e.nextIndex++
			e.blockOpen = true
			e.openTool = false
			e.openIndex = event.Index
			return []*StreamEvent{{Type: EventTypeContentBlockStart, Index: event.Index, ContentBlock: &BlockWire{Type: BlockTypeText, Text: block.Text}}}, nil, nil
		case ir.ToolUseBlock:
			if block.ID == "" {
				return nil, nil, fmt.Errorf("anthropic: ContentBlockStart tool_use id is required")
			}
			if block.Name == "" {
				return nil, nil, fmt.Errorf("anthropic: ContentBlockStart tool_use name is required")
			}
			input, err := streamToolInput(block.Input)
			if err != nil {
				return nil, nil, fmt.Errorf("anthropic: ContentBlockStart tool_use input: %w", err)
			}
			e.nextIndex++
			e.blockOpen = true
			e.openTool = true
			e.openIndex = event.Index
			e.toolInput = string(input)
			e.toolParts = nil
			return []*StreamEvent{{Type: EventTypeContentBlockStart, Index: event.Index, ContentBlock: &BlockWire{
				Type: BlockTypeToolUse, ID: block.ID, Name: block.Name, Input: json.RawMessage("{}"),
			}}}, nil, nil
		default:
			return nil, nil, fmt.Errorf("anthropic: ContentBlockStart carries an unsupported block %T", event.Block)
		}
	case ir.ContentBlockDelta:
		if !e.blockOpen || event.Index != e.openIndex {
			return nil, nil, fmt.Errorf("anthropic: ContentBlockDelta out of grammar order")
		}
		switch delta := event.Delta.(type) {
		case ir.TextDelta:
			if e.openTool {
				return nil, nil, fmt.Errorf("anthropic: TextDelta on tool_use block")
			}
			return []*StreamEvent{{Type: EventTypeContentBlockDelta, Index: event.Index, Delta: &StreamDelta{Type: DeltaTypeTextDelta, Text: delta.Text}}}, nil, nil
		case ir.InputJSONDelta:
			if !e.openTool {
				return nil, nil, fmt.Errorf("anthropic: InputJSONDelta on text block")
			}
			fragment, err := streamInputFragment(delta.PartialJSON)
			if err != nil {
				return nil, nil, fmt.Errorf("anthropic: InputJSONDelta: %w", err)
			}
			e.toolParts = append(e.toolParts, fragment)
			return []*StreamEvent{{Type: EventTypeContentBlockDelta, Index: event.Index, Delta: &StreamDelta{
				Type: DeltaTypeInputJSONDelta, PartialJSON: fragment,
			}}}, nil, nil
		default:
			return nil, nil, fmt.Errorf("anthropic: ContentBlockDelta carries an unsupported delta %T", event.Delta)
		}
	case ir.ContentBlockStop:
		if !e.blockOpen || event.Index != e.openIndex {
			return nil, nil, fmt.Errorf("anthropic: ContentBlockStop out of grammar order")
		}
		if e.openTool {
			if len(e.toolParts) == 0 {
				return e.toolStopWithSynthesizedDelta(event.Index), nil, nil
			}
			if strings.Join(e.toolParts, "") != e.toolInput {
				return nil, nil, fmt.Errorf("anthropic: tool input fragments do not match ToolUseBlock input")
			}
		}
		e.blockOpen = false
		e.openTool = false
		e.toolInput = ""
		e.toolParts = nil
		return []*StreamEvent{{Type: EventTypeContentBlockStop, Index: event.Index}}, nil, nil
	case ir.MessageDelta:
		if !e.started || e.blockOpen || e.deltaSeen {
			return nil, nil, fmt.Errorf("anthropic: MessageDelta out of grammar order")
		}
		reason, seq, err := encodeStopReason(event.StopReason, event.StopSequence)
		if err != nil {
			return nil, nil, err
		}
		e.deltaSeen = true
		return []*StreamEvent{{Type: EventTypeMessageDelta, Delta: &StreamDelta{StopReason: reason, StopSequence: seq}, Usage: &UsageWire{
			InputTokens:  event.Usage.InputTokens,
			OutputTokens: event.Usage.OutputTokens,
		}}}, nil, nil
	case ir.MessageDone:
		if !e.deltaSeen {
			return nil, nil, fmt.Errorf("anthropic: MessageDone out of grammar order")
		}
		e.done = true
		return []*StreamEvent{{Type: EventTypeMessageStop}}, nil, nil
	default:
		return nil, nil, fmt.Errorf("anthropic: unknown event %T", ev)
	}
}

func (e *StreamEncoder) toolStopWithSynthesizedDelta(index int) []*StreamEvent {
	wire := []*StreamEvent{
		{Type: EventTypeContentBlockDelta, Index: index, Delta: &StreamDelta{
			Type: DeltaTypeInputJSONDelta, PartialJSON: e.toolInput,
		}},
		{Type: EventTypeContentBlockStop, Index: index},
	}
	e.blockOpen = false
	e.openTool = false
	e.toolInput = ""
	e.toolParts = nil
	return wire
}

// isIRStringToken reports whether tok is an outer JSON string token, as the
// IR requires for tool inputs and partial fragments. The explicit quote
// check rejects null and other non-string tokens that json.Unmarshal into a
// string would otherwise silently accept as ""; the payload is never
// inspected.
func isIRStringToken(tok json.RawMessage) bool {
	trimmed := strings.TrimSpace(string(tok))
	return trimmed != "" && trimmed[0] == '"'
}

// streamToolInput unwraps an IR tool_use input string token at the streaming
// boundary, rejecting non-string tokens before the existing verbatim unwrap.
func streamToolInput(tok json.RawMessage) (json.RawMessage, error) {
	if !isIRStringToken(tok) {
		return nil, fmt.Errorf("not an IR string token")
	}
	return inputFromIRString(tok)
}

func streamInputFragment(token json.RawMessage) (string, error) {
	if len(token) == 0 {
		return "", fmt.Errorf("partial_json is required")
	}
	if !isIRStringToken(token) {
		return "", fmt.Errorf("partial_json must be an IR string token")
	}
	var fragment string
	if err := json.Unmarshal(token, &fragment); err != nil {
		return "", fmt.Errorf("partial_json must be an IR JSON string token: %w", err)
	}
	return fragment, nil
}

// encodeStopReason maps an IR stop reason (with its conditional stop
// sequence) to the wire stop_reason/stop_sequence pair; it is the inverse of
// the non-streaming decodeStopReason (spec/01 s4).
func encodeStopReason(stop ir.StopReason, stopSequence string) (string, string, error) {
	switch stop {
	case ir.StopEndTurn:
		return "end_turn", "", nil
	case ir.StopMaxTokens:
		return "max_tokens", "", nil
	case ir.StopSequence:
		return "stop_sequence", stopSequence, nil
	case ir.StopToolUse:
		return "tool_use", "", nil
	case ir.StopRefusal:
		return "refusal", "", nil
	default:
		return "", "", fmt.Errorf("anthropic: stop reason %q has no Anthropic equivalent", stop)
	}
}

File: python/src/oxa/anthropic/messages/streamout.py
"""Anthropic Messages streaming encoder: converts IR events into wire stream events (IR → face)."""

from __future__ import annotations

import json
from typing import Any

from oxa.anthropic.messages.constants import (
    BLOCK_TYPE_TEXT,
    BLOCK_TYPE_TOOL_USE,
    DELTA_TYPE_INPUT_JSON_DELTA,
    DELTA_TYPE_TEXT_DELTA,
    EVENT_TYPE_CONTENT_BLOCK_DELTA,
    EVENT_TYPE_CONTENT_BLOCK_START,
    EVENT_TYPE_CONTENT_BLOCK_STOP,
    EVENT_TYPE_MESSAGE_DELTA,
    EVENT_TYPE_MESSAGE_START,
    EVENT_TYPE_MESSAGE_STOP,
    ROLE_ASSISTANT,
    TYPE_MESSAGE,
)
from oxa.anthropic.messages.encode import encode_stop_reason
from oxa.ir import (
    ContentBlockDelta,
    ContentBlockStart,
    ContentBlockStop,
    Event,
    InputJsonDelta,
    Loss,
    MessageDelta,
    MessageDone,
    MessageStart,
    TextBlock,
    TextDelta,
    ToolUseBlock,
)
from oxa.modelmap import Table


class StreamEncoder:
    """Incrementally converts an IR event stream into Anthropic Messages wire events."""

    __slots__ = (
        "_table",
        "_id",
        "_model",
        "_started",
        "_open_index",
        "_open_tool",
        "_tool_input",
        "_tool_parts",
        "_delta_seen",
        "_done",
    )

    def __init__(self, table: Table | None = None) -> None:
        self._table = table
        self._id = ""
        self._model = ""
        self._started = False
        self._open_index = -1
        self._open_tool = False
        self._tool_input = ""
        self._tool_parts: list[str] = []
        self._delta_seen = False
        self._done = False

    def apply(self, ev: Event) -> tuple[list[dict[str, Any]], list[Loss]]:
        """Pushes one IR event and returns the emitted wire events and losses."""
        if self._done:
            raise ValueError("anthropic: event applied after message_done")

        if isinstance(ev, MessageStart):
            if self._started:
                raise ValueError("anthropic: duplicate MessageStart")
            self._started = True
            self._id = ev.id
            self._model = self._table.map(ev.model) if self._table else ev.model
            return [
                {
                    "type": EVENT_TYPE_MESSAGE_START,
                    "message": {
                        "id": self._id,
                        "type": TYPE_MESSAGE,
                        "role": ROLE_ASSISTANT,
                        "model": self._model,
                        "content": [],
                        "stop_reason": None,
                        "usage": {"input_tokens": 0, "output_tokens": 0},
                    },
                }
            ], []

        if isinstance(ev, ContentBlockStart):
            if not self._started or self._open_index >= 0:
                raise ValueError("anthropic: ContentBlockStart out of grammar order")

            if isinstance(ev.block, TextBlock):
                self._open_index = ev.index
                self._open_tool = False
                return [
                    {
                        "type": EVENT_TYPE_CONTENT_BLOCK_START,
                        "index": ev.index,
                        "content_block": {
                            "type": BLOCK_TYPE_TEXT,
                            "text": ev.block.text,
                        },
                    }
                ], []

            if isinstance(ev.block, ToolUseBlock):
                if not ev.block.id or not ev.block.name:
                    raise ValueError("anthropic: ToolUseBlock requires nonempty ID and name")
                self._open_index = ev.index
                self._open_tool = True
                self._tool_input = ev.block.input
                self._tool_parts.clear()
                return [
                    {
                        "type": EVENT_TYPE_CONTENT_BLOCK_START,
                        "index": ev.index,
                        "content_block": {
                            "type": BLOCK_TYPE_TOOL_USE,
                            "id": ev.block.id,
                            "name": ev.block.name,
                            "input": {},
                        },
                    }
                ], []

            raise ValueError("anthropic: ContentBlockStart carries unsupported block")

        if isinstance(ev, ContentBlockDelta):
            if self._open_index != ev.index:
                raise ValueError("anthropic: ContentBlockDelta out of grammar order")

            if self._open_tool:
                if not isinstance(ev.delta, InputJsonDelta):
                    raise ValueError("anthropic: ToolUseBlock received non-input-json delta")
                self._tool_parts.append(ev.delta.partial_json)
                return [
                    {
                        "type": EVENT_TYPE_CONTENT_BLOCK_DELTA,
                        "index": ev.index,
                        "delta": {
                            "type": DELTA_TYPE_INPUT_JSON_DELTA,
                            "partial_json": ev.delta.partial_json,
                        },
                    }
                ], []

            if not isinstance(ev.delta, TextDelta):
                raise ValueError("anthropic: TextBlock received non-text delta")
            return [
                {
                    "type": EVENT_TYPE_CONTENT_BLOCK_DELTA,
                    "index": ev.index,
                    "delta": {
                        "type": DELTA_TYPE_TEXT_DELTA,
                        "text": ev.delta.text,
                    },
                }
            ], []

        if isinstance(ev, ContentBlockStop):
            if self._open_index != ev.index:
                raise ValueError("anthropic: ContentBlockStop out of grammar order")

            events: list[dict[str, Any]] = []
            if self._open_tool:
                if not self._tool_parts:
                    events.append(
                        {
                            "type": EVENT_TYPE_CONTENT_BLOCK_DELTA,
                            "index": ev.index,
                            "delta": {
                                "type": DELTA_TYPE_INPUT_JSON_DELTA,
                                "partial_json": self._tool_input,
                            },
                        }
                    )
                else:
                    if "".join(self._tool_parts) != self._tool_input:
                        raise ValueError(
                            "anthropic: ToolUseBlock input does not equal concatenated InputJSONDelta fragments"
                        )
                self._open_tool = False
                self._tool_input = ""
                self._tool_parts.clear()

            self._open_index = -1
            events.append(
                {
                    "type": EVENT_TYPE_CONTENT_BLOCK_STOP,
                    "index": ev.index,
                }
            )
            return events, []

        if isinstance(ev, MessageDelta):
            if not self._started or self._open_index >= 0:
                raise ValueError("anthropic: MessageDelta out of grammar order")
            reason, seq = encode_stop_reason(ev.stop_reason, ev.stop_sequence)
            self._delta_seen = True
            delta_dict: dict[str, Any] = {"stop_reason": reason}
            if seq is not None:
                delta_dict["stop_sequence"] = seq
            return [
                {
                    "type": EVENT_TYPE_MESSAGE_DELTA,
                    "delta": delta_dict,
                    "usage": {
                        "input_tokens": ev.usage.input_tokens,
                        "output_tokens": ev.usage.output_tokens,
                    },
                }
            ], []

        if isinstance(ev, MessageDone):
            if not self._delta_seen:
                raise ValueError("anthropic: MessageDone out of grammar order")
            self._done = True
            return [{"type": EVENT_TYPE_MESSAGE_STOP}], []

        raise ValueError(f"anthropic: unknown event type {type(ev)}")
